src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -524,10 +630,389 @@ GetRelationAccessMode(Oid relationId, ShardPlacementAccessType accessType)
+									  "executed in the same transaction"),
+							errhint("Try re-running the transaction with "
+									"\"SET LOCAL citus.multi_shard_modify_mode TO "
+									"\'sequential\';\"")));
+		}
+		else
+		{
+			ereport(DEBUG1, (errmsg("switching to sequential query execution mode"),
+							 errdetail(
+								 "Reference relation \"%s\" is modified, which might lead "
+								 "to data inconsistencies or distributed deadlocks via "
+								 "parallel accesses to hash distributed relations due to "
+								 "foreign keys. Any parallel modification to "
+								 "those hash distributed relations in the same "
+								 "transaction can only be executed in sequential query "
+								 "execution mode", relationName)));
+
+			/*
+			 * Switching to sequential mode is admittedly confusing and, could be useless
+			 * and less performant in some cases. However, if we do not switch to
+			 * sequential mode at this point, we'd loose the opportunity to do so
+			 * later when a parallel query is executed on the hash distributed relations
+			 * that are referencing this reference table.
+			 */
+			SetLocalMultiShardModifyModeToSequential();
+		}
+	}
+}
+
+
+/*
+ * CheckConflictingParallelRelationAccesses is mostly a wrapper around
+ * HoldsConflictingLockWithReferencedRelations().  We're only interested in parallel
+ * accesses to distributed tables that refers reference tables via foreign constraint.
+ *
+ */
+static void
+CheckConflictingParallelRelationAccesses(Oid relationId, ShardPlacementAccessType
+										 accessType)
+{
+	DistTableCacheEntry *cacheEntry = NULL;
+	Oid conflictingReferencingRelationId = InvalidOid;
+	ShardPlacementAccessType conflictingAccessType = PLACEMENT_ACCESS_SELECT;
+
+	if (!EnforceForeignKeyRestrictions || !IsDistributedTable(relationId))
+	{
+		return;
+	}
+
+	cacheEntry = DistributedTableCacheEntry(relationId);
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+		  cacheEntry->referencedRelationsViaForeignKey != NIL))
+	{
+		return;
+	}
+
+	if (MultiShardConnectionType == PARALLEL_CONNECTION &&
+		HoldsConflictingLockWithReferencedRelations(relationId,
+													&conflictingReferencingRelationId,
+													&conflictingAccessType))
+	{
+		char *relationName = get_rel_name(relationId);
+		char *conflictingRelationName = get_rel_name(conflictingReferencingRelationId);
+
+		char *accessTypeText = PlacementAccessTypeToText(accessType);
+		char *conflictingAccessTypeText =
+			PlacementAccessTypeToText(conflictingAccessType);
+
+		if (ParallelQueryExecutedInTransaction())
+		{
+			/*
+			 * If there has already been a parallel query executed, the sequential mode
+			 * would still use the already opened parallel connections to the workers,
+			 * thus contradicting our purpose of using sequential mode.
+			 */
+			ereport(ERROR, (errmsg("cannot execute parallel %s on relation \"%s\" "
+								   "since it holds conflicting locks on reference "
```

Comment:
I'd omit the conflicting locks bit, since this is a fairly deep down implementation detail of postgres and citus, and instead be more precise about what the user did that led to this situation:

e.g.
```
ERROR: cannot execute parallel %s on table \"%s\" after a %s command on reference table \"%s\" because there is a foreign key between them and the operation cannot be serialised
```

Reply:
I think I'll prefer `has been modified in this transaction` over `the operation cannot be serialized` as you've suggested above. Because that sounds easier to understand for anyone

Reply:
note that my comments were about different messages. Here I think we don't know that the relation has been modified? (could be DDL after SELECT)